### PR TITLE
Chapter1 tests

### PR DESCRIPTION
**Pull request by zulzu**
add tests for pure functions (add / subtract / multiply / division / average / modulus )

### CODE REVIEW

File: Chapter1/Tests/CalculatorTests/AddTests.swift [hunk 1]
```diff
@@ -2,38 +2,55 @@ import XCTest
...
+  }
+  
+  func testAdd_whenFirstIsZero_returnsSecond() throws {
+    // neutral element: 0 + x = x
+    let first = 0
+    let second = 5
+    
+    let result = Calculator.add(first, second)
+    
+    XCTAssertEqual(result, 5)
```
Comment (cipolleschi) [inline Chapter1/Tests/CalculatorTests/AddTests.swift]: A better way to write this test is to use second as the expected result.
In this way, it's explicit that when the first element is zero, the expected result is the second element, independently from its value.

It also avoid using twice a value (DRY principle). What if we change the value of the `second` variable and we forget to update the expectation?
```suggestion
    XCTAssertEqual(result, second)
```

File: Chapter1/Tests/CalculatorTests/AddTests.swift [hunk 2]
```diff
@@ -2,38 +2,55 @@ import XCTest
...
+  }
+  
+  func testAdd_whenSecondIsZero_returnsFirst() throws {
+    // neutral element: x + 0 = x
+    let first = 3
+    let second = 0
+    
+    let result = Calculator.add(first, second)
+    
+    XCTAssertEqual(result, 3)
```
Comment (cipolleschi) [inline Chapter1/Tests/CalculatorTests/AddTests.swift]: Same reasoning as above.
```suggestion
    XCTAssertEqual(result, first)
```

File: Chapter1/Tests/CalculatorTests/AverageTests.swift [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+    XCTAssertEqual(result, 5)
+  }
+  
+  func testAverage_whenEmptyArray_returnsNil() throws {
+    let values : [Int] = []
+    
+    let result = Calculator.average(values)
+    
+    XCTAssertEqual(result, nil)
+  }
```
Comment (cipolleschi) [inline Chapter1/Tests/CalculatorTests/AverageTests.swift:20]: What happens when we have a single value?
What happens with all negative values? And with some negative values?
What happens with values that are all the same?

When writing tests, we should always consider boundaries and special conditions. In this case, we need to test also the special conditions above.

File: Chapter1/Tests/CalculatorTests/DivisionTests.swift [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+    XCTAssertEqual(result, 1)
+  }
+  
+  func testDivision_whenDividingByZero_returnsNil() throws {
+    let first = 3
+    let second = 0
+    
+    let result = try? Calculator.division(first, second)
+    
+    XCTAssertEqual(result, nil)
```
Comment (cipolleschi) [inline Chapter1/Tests/CalculatorTests/DivisionTests.swift]: You testes very specific examples. This division can pass both tests if implemented like this:
```swift
func division(_ first: Int, _ second: Int) -> Int? {
  guard second != 0 else { return nil }
  return 1
}
```

You should add tests for generic values. You should add tests with negative values.
You should add tests for the [divisions properties](https://www.math-only-math.com/Properties-of-Division.html).
You should add a test for non-integer division (and decide whether it has to return a double or the rounded down value) and a test for what happens when the numerator (first) is 0 (an important special case)

File: Chapter1/Tests/CalculatorTests/MultiplyTests.swift [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+    let result2 = Calculator.multiply(second, first)
+    
+    XCTAssertEqual(result1, result2)
+  }
+  
+  func testMultiply_twoMultipliedByThree_MultipliedByFour_AndTwoMultipliedBy_ThreeMultipliedByFour_returnsTheSameResult() throws {
+    // associative property: (a * b) * c = a * (b * c)
+    let first = 2
+    let second = 3
+    let third = 4
```
Comment (cipolleschi) [inline Chapter1/Tests/CalculatorTests/MultiplyTests.swift]: as a rule of thumbs, when you can use various values, try not to reuse them. For example, in the previous test, we run `2*3` and `3*2`. Try not to reuse these values in the following tests: by doing so we are going to cover more use cases in our tests in a simple way.

For example, you could have used 5, 19 and 54. 

We are also missing some special cases: what happens when the numbers we use are negative? 
Multiplication has an interesting property to test, using the modulus operation we tested above.

**Review by cipolleschi — CHANGES_REQUESTED** [review]
Hi Andras! Thank you so much for the correction on the readme.
So far, so good. Great starting point.

I left some comments to raise a bit of awareness about testing special cases/conditions and boundaries.

There are a few tests we can add!

File: Chapter1/Tests/CalculatorTests/DivisionTests.swift [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+    let result = try Calculator.division(first, second)
+    
+    XCTAssertEqual(result, 1)
+  }
+  
+  func testDivision_whenDividingByZero_returnsNil() throws {
+    let first = 3
+    let second = 0
+    
+    let result = try? Calculator.division(first, second)
```
Comment (cipolleschi) [inline Chapter1/Tests/CalculatorTests/DivisionTests.swift]: This is not the best approach. We are not checking that an error is raised when we are dividing by zero. Also, we do not test which kind of error is thrown.

This test will work well also if the `division` signature would be 
```swift
func division(_ num: Int, _ den: Int) -> Int?
```
which is not.

When a function declares that it `throws`, we want to test that an error is thrown in those conditions and which actual error is thrown.

You can use the [XCTAssertThrowsError](https://developer.apple.com/documentation/xctest/error_assertions) function to verify that.

File: Chapter1/Tests/CalculatorTests/DivisionTests.swift [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+    var thrownError: Error?
+    let first = 3
+    let second = 0
+        
+    XCTAssertThrowsError(try Calculator.division(first, second)) {
+      thrownError = $0
+    }
+    XCTAssertTrue(thrownError is Calculator.CalculatorError)
+    XCTAssertEqual(thrownError as? Calculator.CalculatorError, .divisionByZero)
+  }
```
Comment (cipolleschi) [inline Chapter1/Tests/CalculatorTests/DivisionTests.swift:24]: This is a good approach. Another alternative is to assert the error directly in the closure:

```swift
XCTAssertThrowsError(try Calculator.division(first, second)) {
  guard let calcError = try? XCTUnwrap($0 as Calculator.CalculatorError) else {
    XCTFail("Error is not a Calculator error but a \(type(of: $0))")
    return
  }
  XCTAssertEqual(calcError, .divisionByZero
}
```

**Review by cipolleschi — APPROVED** [review]
Very nice! 👏 
I left a comment that is more for your knowledge. You don't need to fix this before moving to the next chap!
